web/src/main/java/com/bazaarvoice/emodb/web/migrator/LocalRangeMigrator.java

```diff
@@ -0,0 +1,488 @@
+                WAIT_FOR_ALL_TRANSFERS_COMPLETE_CHECK_INTERVAL, WAIT_FOR_ALL_TRANSFERS_COMPLETE_TIMEOUT);
+    }
+
+    public LocalRangeMigrator(MigratorTools migratorTools, MigratorWriterFactory writerFactory, LifeCycleRegistry lifecycle, final MetricRegistry metricRegistry, int threadCount,
+                              int batchSize, Duration waitForAllTransfersCompleteCheckInterval, Duration waitForAllTransfersCompleteTimeout) {
+        _migratorTools = migratorTools;
+        _writerFactory = writerFactory;
+        _threadCount = threadCount;
+        _batchSize = batchSize;
+        _waitForAllTransfersCompleteCheckInterval = waitForAllTransfersCompleteCheckInterval;
+        _waitForAllTransfersCompleteTimeout = waitForAllTransfersCompleteTimeout;
+
+        _activeRangeMigrations = metricRegistry.counter(MetricRegistry.name("bv.emodb.migrator", "Migrator", "active-range-migrations"));
+        _blockedRangeMigrations = metricRegistry.counter(MetricRegistry.name("bv.emodb.migrator", "Migrator", "blocked-range-migrations"));
+        _activeBatches = metricRegistry.counter(MetricRegistry.name("bv.emodb.migrator", "Migrator", "active-batches"));
+        _batchRowsSubmitted = metricRegistry.counter(MetricRegistry.name("bv.emodb.migrator", "Migrator", "batch-rows-submitted"));
+        _batchesSubmitted = metricRegistry.counter(MetricRegistry.name("bv.emodb.migrator", "Migrator", "batches-submitted"));
+        _waitingForBatchesComplete = metricRegistry.counter(MetricRegistry.name("bv.emodb.migrator", "Migrator", "waiting-for-batches-complete"));
+        _failedRangeMigrations = metricRegistry.meter(MetricRegistry.name("bv.emodb.migrator", "Migrator", "failed-range-migrations"));
+
+        _rawBytesMigrated = CacheBuilder.newBuilder()
+                .build(new CacheLoader<String, Counter>() {
+                    @Override
+                    public Counter load(String placement)
+                            throws Exception {
+                        return metricRegistry.counter(MetricRegistry.name("bv.emodb.migrator.Migrator.placement", placement, "raw-bytes-migrated"));
+                    }
+                });
+
+        lifecycle.manage(this);
+    }
+
+    @Override
+    public void start() throws Exception {
+        _timeoutService = Executors.newScheduledThreadPool(1, new ThreadFactoryBuilder().setNameFormat("range-migrator-timeout-%d").build());
+        _shutdown = false;
+    }
+
+    @Override
+    public void stop() throws Exception {
+        _shutdown = true;
+        _timeoutService.shutdownNow();
+        synchronized (_batchServices) {
+            for (ExecutorService service : _batchServices) {
+                service.shutdownNow();
+            }
+            _batchServices.clear();
+        }
+    }
+
+    public RangeScanUploaderResult migrate(final int taskId, ScanOptions options, final String placement, ScanRange range, int maxConcurrentWrites) throws IOException, InterruptedException {
```

Comment:
I know this logic is largely copy-pasta'd from stash, but in this case I think it may be real complexity overkill.  In stash's case the workflow consists of numerous steps with varying execution requirements -- for example, the time to read raw records from Cassandra != time to resolve those records into JSON != time to transport JSON to S3 -- so the pipeline and multiple execution threads help move that through quickly.  Here, however, there's a simple near 1-to-1 relationship in terms of execution time and requirements for reading unblocked rows and writing those same rows back to the blocked tables.  It may be that a simple loop over the results which writes them back blocked is all that's necessary.